doc/crawler.rst

```diff
@@ -0,0 +1,91 @@
+Crawler -- mm2_crawler
+======================
+
+Overview
+--------
+
+The crawler scans all available mirrors if their content is up to date. The
+crawler only scans mirrors which are not disabled (host.user_active,
+host.admin_active, site.user_active, site.admin_active).
+
+The crawler retrieves the list of active mirrors from the database and
+spawns a thread for each host. The crawling of a host consists of a loop
+over all categories which have been configured for this host and one category
+is crawled after another.
+
+The crawler has the possibility to crawl a category via HTTP, FTP or RSYNC.
+For each category the protocol to crawl is selected from the available
+category protocols. RSYNC has the highest priority and is followed by HTTP
+and the protocol used last (if no other protocol is available) is FTP.
+
+After the crawl has finished (successful or unsuccessful) the duration
+and the timestamp of the crawl is recorded in the host. After the crawl of
+each category the files which are up to date in the category are stored
+in the database. Crawl failures (timeouts, network problems) usually result
+in a complete 'marked-as-not-up-to-date' of that mirror. Multiple consecutive
+crawl failures (default 4) will disable the host completely (host.user_active).
+
+The crawler requires enormous amounts of memory and for 40 threads crawling
+mirrors in parallel at least 32GB of memory are required. At the end of
+each crawl thread the garbage collector is explicitly called in the hope
+that some unused memory is freed again.
+
+Crawling protocol
+-----------------
+
+As previously mentioned the crawler uses either RSYNC, HTTP or FTP to crawl
+a connected mirror. As also already mentioned the crawler spawns a thread
+for each host it has to crawl and then starts crawling one category configured
+after another. For each category the protocol decision made anew.
+
+It the category supports RSYNC the whole category is scanned using RSYNC with
+a single network connection. If it was able to find all files the category
+is marked as up to date and the next category follows. If no RSYNC URL is
+available the crawler uses FTP or HTTP. FTP requires one network connection
+per directory and using HTTP each file is crawled separately. Depending on
+the configuration of the remote host this can mean one network connection
+per file or, if the remote host supports HTTP keep-alive, one connection
+for multiple files or the whole category. This depends again on the
+configuration of the remote host.
+
+Using HTTP most of the files are only read via HTTP HEAD. Not the actual data
+of the file is downloaded. Only the metadata is downloaded. Only files with
+the name 'repomd.xml' are actually completed downloaded and their SHA256 sum
+is compared to the one in the database.
```

Comment:
Looking at the code of the crawler, atomic is not mentioned at all. This is probably one of the reasons crawling takes so long. The atomic tree needs to be handled differently.